[FEAT] 캘린더 월 별 공부 기록 조회 기능 추가

qbobl5 opened the pull request:
# 📄 Work Description
- 캘린더 월 별 공부 기록 조회 기능 추가

# ⚙️ ISSUE
- closed #23

# 📷 Screenshot
- 포스트맨
<img width="1015" alt="스크린샷 2024-08-12 오후 2 29 07" src="https://github.com/user-attachments/assets/56d38a22-b5a8-4e2e-bf17-ef71e23697f1">

<details>
<summary>전체 결과 보기</summary>

```json
{
    "year": 2024,
    "month": 8,
    "daily": [
        {
            "day": 2,
            "time": 20000,
            "subjects": [
                {
                    "id": 4,
                    "name": "JS"
                }
            ]
        },
        {
            "day": 10,
            "time": 20000,
            "subjects": [
                {
                    "id": 6,
                    "name": "CSS"
                }
            ]
        },
        {
            "day": 12,
            "time": 50000,
            "subjects": [
                {
                    "id": 1,
                    "name": "HTML"
                }
            ]
        },
        {
            "day": 12,
            "time": 20000,
            "subjects": [
                {
                    "id": 4,
                    "name": "JS"
                }
            ]
        }
    ]
}
```
</details>


# 💬 To Reviewers
일단 native query 를 사용했습니다. `month()` 와 `year()` 함수가 jpa 로 사용이 불가하다고 알고 있어서.. 🥲
혹시 더 깔끔하게 수정할 수 있는 방법을 아시면 추천 부탁드립니다.

qbobl5 committed:
[feat/#22]: service interface 생성

qbobl5 committed:
[feat/#22]: service interface implements 추가

qbobl5 committed:
[refactor/#22]: study record service 사용하는 부분 리팩토링

qbobl5 committed:
[refactor/#22]: transactional (readonly) 위치 변경

qbobl5 committed:
Merge branch 'refactor/#22-study-record'

qbobl5 committed:
[refactor/#23]: 선언 변경
- impl 받은 class 대신 interface 로 변경

qbobl5 committed:
[refactor/#23]: ShowTimerResponse 경로 변경
- studyrecord -> timer 로 경로 변경
- 리팩토링

qbobl5 committed:
[feat/#23]: 일 별 공부기록 response 추가
- day (일 ex. 2)
- time (공부한 시간(초) ex. 36000)
- subjects (공부한 과목 목록)

qbobl5 committed:
[feat/#23]: 특정 월 공부기록 response 추가
- year (연 ex.2024)
- month (월 ex.8)
- daily (일 별 공부기록 response 목록)

qbobl5 committed:
[fix/#23]: Collect 대신 toList 로 변경

qbobl5 committed:
[feat/#23]: 연도와 월의 공부 기록 목록 조회

qbobl5 committed:
[feat/#23]: service interface 에 해당하는 월 공부 목록을 조회하는 기능 추가

qbobl5 committed:
[feat/#23]: 해당하는 월 공부 목록을 조회하는 기능 추가
- 회원 존재 여부 검사
- response dto 반환

qbobl5 committed:
[feat/#23]: controller 에 getMonthlyRecords 추가

qbobl5 committed:
[refactor/#23]: 변수명 serviceImpl -> service 로 리팩토링

qbobl5 committed:
[fix/#23]: readonly 제거

qbobl5 committed:
[fix/#23]: 필요없는 import 문 삭제

qbobl5 committed:
[test/#23]: 해당하는 월 공부 목록 조회 test 추가
- 연도와 월 통해 공부 기록을 조회하는 테스트
- 기록이 없는 월을 조회하는 테스트
  - 기록이 없으면 null 이 아니라 [] 반환

qbobl5 committed:
[fix/#23]: @RequestParam 추가

qbobl5 committed:
[test/#23]: 컨트롤러 테스트 추가
- 정상적으로 연도와 월을 받은 경우 (200)
- 월이 전달되지 않은 경우 (400)
- 연도가 전달되지 않은 경우 (400)

qbobl5 committed:
[fix/#23]: confilct 해결

qbobl5 committed:
[fix/#23]: confilct 해결

qbobl5 committed:
[fix/#23]: 잘못된 query 수정

qbobl5 committed:
[fix/#23]: statusCode 직접 쓰지 않고 value 가져오는 것으로 변경

Commit:
format: Google Java Format

khwoowoo commented on src/main/java/com/woohaengshi/backend/repository/StudyRecordRepository.java:
```diff
@@ -3,10 +3,29 @@
+                            + "  AND r.member_id = :memberId\n"
+                            + "ORDER BY r.date ASC;",
+            nativeQuery = true)
+    List<Object[]> findByYearAndMonthAndMemberId(
```
개인적으로 궁금한데 이런 상황에서는 DTO를 생성하는 것보다는 Object로 받는 게 낫나요??

qbobl5 replied:
전체 SELECT 해오는 것들이 복잡하지 않은 로직으로 되어있으면 DTO 생성해서 JPQL 사용할 수 있습니다.
그런데 해당 쿼리에서는 중복되는 부분 (date, time) 이 있으니까 그건 전달될 dto의 변수로 빼고 나머지 subject 를 list 로 해서 새로운 dto에 넣고, 또 이 subjectList 를 전달될 dto의 변수로 넣어주는 복잡한 과정이 있어서 native query를 사용했습니다 🫠

참고로 제걸 jpql 로 하려면 또다른 dto 를 만들어서 그 dto 로 받은 다음 해당 dto 의 내용을 제가 전달한 dto 들에게 또 전달해줘야 하는 과정이 생겨서.. (List<Subject~~> 때문에) 전 native query 로 했습니다!

qbobl5 replied:
아 그리고 저는 매칭할 dto 가 존재하지 않아서 일단 List<Object[]> 타입으로 다 받아왔습니다

qbobl5 committed:
[fix/#23]: year, month 대신 YearMonth 타입으로 한번에 날짜 받기
[ex]
- ?year=2024&month=8 (X)
- ?date=2024-08 (O)

이유: LocalDate 는 따로 컨트롤러에서 2024-08 을 받고 작업 필요. 그래서 YearMonth 로 한번에 받고 처리

qbobl5 committed:
Merge branch 'feat/#23-monthly-calendar of https://github.com/woohaengshi/backend into feat/#23-monthly-calendar

rlfrkdms1 commented on src/main/java/com/woohaengshi/backend/dto/response/studyrecord/ShowMonthlyRecordResponse.java:
```diff
@@ -0,0 +1,38 @@
+                        .map(
+                                record ->
+                                        ShowDailyRecordResponse.of(
+                                                ((Number) record[0]).intValue(),
+                                                ((Number) record[1]).intValue(),
+                                                List.of(
+                                                        Subject.builder()
+                                                                .id((Long) record[2])
+                                                                .name((String) record[3])
+                                                                .build())))
+                        .toList());
```
이게 모두 ShowDailyResponse로 들어가는게 어떤가요?

rlfrkdms1 replied:
그리고 지금 subject각각의 id와 이름을 Subject로 만들고 이걸 다시 까서 dto로 만드는 과정을 거치고 있는데, Subject로 만드는 과정은 없애는게 낫지 않을까요?

ShowSubjectResponse.from을 재사용하기 위해 그런거라면 from을 사용하지 마시고, of를 만들거나 그냥 생성자를 사용하시는게 더 좋아보입니다.

qbobl5 replied:
굳굳 안그래도 고민했던 부분인데 수정해보겠습니다

rlfrkdms1 commented on src/main/java/com/woohaengshi/backend/service/studyrecord/StudyRecordServiceImpl.java:
```diff
@@ -41,6 +43,19 @@ public void save(SaveRecordRequest request, Long memberId) {
+        int month = date.getMonthValue();
+
+        return ShowMonthlyRecordResponse.of(
+                year,
+                month,
```
그냥 date를 넘겨줘도 되지 않을까요?

qbobl5 replied:
흠 이건 프론트에서 맡은 분이랑 이야기해봤는데 다시 한 번 조정해보도록 할게욥~!! 일단 year, month 따로 주는 걸로 이야기를 했어서 date 로 한번에 주는게 편하다고 하시면 바꾸도록 하겠슴당

qbobl5 replied:
프론트에서 따로 주는게 더 좋다고 해서 ..... ^--^..... 이것은 그냥 year, month 로 주도록 해도 될까요오오옹??

rlfrkdms1 commented on src/test/java/com/woohaengshi/backend/controller/StudyRecordControllerTest.java:
controller 테스트는 아직 수정되지 않은 것 같습니다. 

참고로 `YearMonth.now()`사용하면 좋을 것 같습니다.

qbobl5 replied:
아 맞네요.... 감사합니다~! 테스트 코드 다시 작성해보도록 하겠습니당

rlfrkdms1 commented:
Repository Test를 작성하셨다고 말씀하셨는데, 작성되어있지 않네요.
native query를 사용하셨으니 Repository Test 작성은 필수적이라고 생각합니다 ! 
작성하셔서 같이 올려주시면 더 좋을 것 같아요

qbobl5 committed:
[test/#23]: repository test 추가
- 공부 기록이 일 별로 제대로 조회가 되는지 확인
  - index [2] 를 비교하지 않은 이유는 해당 값이 id 라 실제로 예측할 수 없는 값이므로 따로 비교하지 않았습니다.

Commit:
format: Google Java Format

qbobl5 committed:
[test/#23]: controller test 수정

qbobl5 committed:
[test/#23]: repository test 수정
- beforeEach 제거
- 저장 method 추가
- 1월과 2월의 공부 기록을 저장 후, 1월의 기록을 불러올 때 잘 불러와지는지 확인하는 테스트 추가

qbobl5 committed:
[fix/#23]: dto 생성 부분 수정
- service 에서 dto 에 year, month 따로 전달하지 않고 date 만 전달
- ShowDailyRecordResponse 에서 subject 생성하지 않고 Monthly 에서 바로 생성하도록 변경

qbobl5 committed:
[fix/#23] confilct 해결

qbobl5 committed:
Merge branch 'dev' into feat/#23-monthly-calendar

Commit:
format: Google Java Format

khwoowoo reviewed (approved):
수고많으셧습니다 👍
